File: demos/53-askdocs/api/events.py
"""Forge Events publish client for AskDocs (epic 53.02)."""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any


@dataclass
class EventsConfig:
    base_url: str
    source: str
    subject: str


def load_events_config(environ: dict[str, str] | None = None) -> EventsConfig:
    env = environ if environ is not None else os.environ
    base = (env.get("FORGE_EVENTS_URL") or "").strip() or "http://host.docker.internal:4105"
    source = (env.get("FORGE_SERVICE_NAME") or "").strip() or "askdocs-api"
    subject = (env.get("FORGE_EVENTS_SUBJECT") or "").strip() or "document.uploaded"
    return EventsConfig(base_url=base.rstrip("/"), source=source, subject=subject)


class EventsClient:
    def __init__(self, cfg: EventsConfig | None = None) -> None:
        self.cfg = cfg or load_events_config()

    def _request(
        self,
        method: str,
        url: str,
        data: bytes | None = None,
        headers: dict[str, str] | None = None,
        timeout: float = 15.0,
    ) -> tuple[int, bytes]:
        req = urllib.request.Request(url, data=data, method=method, headers=headers or {})
        try:
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                return int(resp.status), resp.read()
        except urllib.error.HTTPError as exc:
            body = exc.read() if exc.fp else b""
            return int(exc.code), body

    def ping(self) -> None:
        code, body = self._request("GET", f"{self.cfg.base_url}/health/ready", timeout=10.0)
        if code != 200:
            raise RuntimeError(f"events ready HTTP {code}: {body[:256]!r}")

    def publish_document_uploaded(
        self,
        document_id: str,
        object_key: str,
        content_type: str,
        title: str = "",
        uploaded_at: str | None = None,
    ) -> None:
        uploaded_at = uploaded_at or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
        data: dict[str, Any] = {
            "document_id": document_id,
            "object_key": object_key,
            "content_type": content_type,
            "uploaded_at": uploaded_at,
            "source": self.cfg.source,
        }
        if title:
            data["title"] = title
        payload = {
            "subject": self.cfg.subject,
            "source": self.cfg.source,
            "data": data,
        }
        raw = json.dumps(payload).encode()
        headers = {
            "Content-Type": "application/json",
            "Idempotency-Key": document_id,
        }
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/events",
            data=raw,
            headers=headers,
            timeout=15.0,
        )
        if code != 202:
            raise RuntimeError(f"events publish HTTP {code}: {body[:1024]!r}")

File: demos/52-snapnote/api/events.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

type eventsConfig struct {
	BaseURL string
	Source  string
	Subject string
}

func loadEventsConfig() eventsConfig {
	base := strings.TrimSpace(os.Getenv("FORGE_EVENTS_URL"))
	if base == "" {
		base = "http://host.docker.internal:4105"
	}
	source := strings.TrimSpace(os.Getenv("FORGE_SERVICE_NAME"))
	if source == "" {
		source = "snapnote-api"
	}
	subject := strings.TrimSpace(os.Getenv("FORGE_EVENTS_SUBJECT"))
	if subject == "" {
		subject = "attachment.uploaded"
	}
	return eventsConfig{
		BaseURL: strings.TrimRight(base, "/"),
		Source:  source,
		Subject: subject,
	}
}

type eventsClient struct {
	cfg        eventsConfig
	httpClient *http.Client
}

func newEventsClient(cfg eventsConfig) *eventsClient {
	return &eventsClient{
		cfg: cfg,
		httpClient: &http.Client{
			Timeout: 15 * time.Second,
		},
	}
}

func (c *eventsClient) enabled() bool {
	return c != nil && c.cfg.BaseURL != "" && c.cfg.Subject != ""
}

func (c *eventsClient) Ping(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.cfg.BaseURL+"/health/ready", nil)
	if err != nil {
		return err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("events ready HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}

type attachmentUploadedData struct {
	AttachmentID string `json:"attachment_id"`
	NoteID       string `json:"note_id"`
	ObjectKey    string `json:"object_key"`
	ContentType  string `json:"content_type"`
	UploadedAt   string `json:"uploaded_at"`
	Source       string `json:"source,omitempty"`
}

// PublishAttachmentUploaded publishes attachment.uploaded with Idempotency-Key=attachment_id.
func (c *eventsClient) PublishAttachmentUploaded(ctx context.Context, att *Attachment) error {
	if !c.enabled() {
		return fmt.Errorf("events url not configured")
	}
	body := map[string]any{
		"subject": c.cfg.Subject,
		"source":  c.cfg.Source,
		"data": attachmentUploadedData{
			AttachmentID: att.ID,
			NoteID:       att.NoteID,
			ObjectKey:    att.ObjectKey,
			ContentType:  att.ContentType,
			UploadedAt:   time.Now().UTC().Format(time.RFC3339),
			Source:       c.cfg.Source,
		},
	}
	raw, err := json.Marshal(body)
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/events", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	// Business key for publish dedup + consumer idempotency correlation.
	req.Header.Set("Idempotency-Key", att.ID)
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode != http.StatusAccepted {
		return fmt.Errorf("events publish HTTP %d: %s", resp.StatusCode, string(respBody))
	}
	return nil
}
